test/ability_dry_skin.c

```diff
@@ -0,0 +1,114 @@
+    } WHEN {
+        TURN { MOVE(player, MOVE_SUNNY_DAY); }
+    } SCENE {
+        ABILITY_POPUP(player, ABILITY_DRY_SKIN);
+        HP_BAR(player, damage: TEST_MAX_HP / 8);
+        MESSAGE("The Heliolisk's Dry Skin takes its toll!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Dry Skin heals 1/8th Max HP in Rain")
+{
+    GIVEN {
+        PLAYER(SPECIES_HELIOLISK) { Ability(ABILITY_DRY_SKIN); HP(100); MaxHP(TEST_MAX_HP); };
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_RAIN_DANCE); }
+    } SCENE {
+        ABILITY_POPUP(player, ABILITY_DRY_SKIN);
+        MESSAGE("Heliolisk's Dry Skin restored its HP a little!");
+        HP_BAR(player, hp: TEST_MAX_HP / 8 + 100);
+    }
+}
+
+SINGLE_BATTLE_TEST("Dry Skin increases damage taken from Fire-type moves by 25%", s16 damage)
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_EMBER].type == TYPE_FIRE);
+        PLAYER(SPECIES_WOBBUFFET);
+        OPPONENT(SPECIES_HELIOLISK) { Ability(ABILITY_DRY_SKIN); };
+    } WHEN {
+        TURN {MOVE(player, MOVE_EMBER); }
+    } SCENE {
+        MESSAGE("Wobbuffet used Ember!");
+        HP_BAR(opponent, captureDamage: &results[i].damage);
+    } FINALLY {
+        EXPECT_MUL_EQ(results[0].damage, Q_4_12(1.25), results[1].damage);
+    }
+}
+
+SINGLE_BATTLE_TEST("Dry Skin heals 25% when hit by water type moves")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_BUBBLE].type == TYPE_WATER);
+        PLAYER(SPECIES_HELIOLISK) { Ability(ABILITY_DRY_SKIN); HP(100); MaxHP(TEST_MAX_HP); };
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_BUBBLE); }
+    } SCENE {
+        ABILITY_POPUP(player, ABILITY_DRY_SKIN);
+        HP_BAR(player, hp: TEST_MAX_HP / 4 + 100);
+        MESSAGE("Heliolisk restored HP using its Dry Skin!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Dry Skin does not activate if protected")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_BUBBLE].type == TYPE_WATER);
+        PLAYER(SPECIES_HELIOLISK) { Ability(ABILITY_DRY_SKIN); HP(100); MaxHP(TEST_MAX_HP); };
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_PROTECT); MOVE(opponent, MOVE_BUBBLE); }
+    } SCENE {
+        NONE_OF { ABILITY_POPUP(player, ABILITY_DRY_SKIN); HP_BAR(player); MESSAGE("Heliolisk restored HP using its Dry Skin!"); }
+    }
+}
+
+SINGLE_BATTLE_TEST("Dry Skin is only triggered once on multi strike moves")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_WATER_SHURIKEN].type == TYPE_WATER);
+        ASSUME(gBattleMoves[MOVE_WATER_SHURIKEN].effect == EFFECT_MULTI_HIT);
+        PLAYER(SPECIES_HELIOLISK) { Ability(ABILITY_DRY_SKIN); HP(100); MaxHP(TEST_MAX_HP); };
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_WATER_SHURIKEN); }
+    } SCENE {
+        ABILITY_POPUP(player, ABILITY_DRY_SKIN);
+        HP_BAR(player, hp: TEST_MAX_HP / 4 + 100);
+        MESSAGE("Heliolisk restored HP using its Dry Skin!");
+    }
```

Comment:
I would put a FINALLY to check the final health at the end of the turn. Afaik, it's currently only checking if the HP Bar only restoring once.

Reply:
This effect is functionally identical to water absorb, whose test isn't using `FINALLY`.

I could add that if desired, but wanna make sure it's actually necessary given this context.

```c
SINGLE_BATTLE_TEST("Water Absorb heals 25% when hit by water type moves")
{
    GIVEN {
        ASSUME(gBattleMoves[MOVE_BUBBLE].type == TYPE_WATER);
        PLAYER(SPECIES_POLIWAG) { Ability(ABILITY_WATER_ABSORB); HP(1); MaxHP(TEST_MAX_HP); };
        OPPONENT(SPECIES_WOBBUFFET);
    } WHEN {
        TURN { MOVE(opponent, MOVE_BUBBLE); }
    } SCENE {
        ABILITY_POPUP(player, ABILITY_WATER_ABSORB);
        HP_BAR(player, hp: TEST_MAX_HP / 4 + 1);
        MESSAGE("Poliwag restored HP using its Water Absorb!");
    }
}
```

Neither does Volt Absorb.

```c
SINGLE_BATTLE_TEST("Volt Absorb heals 25% when hit by electric type moves")
{
    GIVEN {
        ASSUME(gBattleMoves[MOVE_THUNDER_SHOCK].type == TYPE_ELECTRIC);
        PLAYER(SPECIES_JOLTEON) { Ability(ABILITY_VOLT_ABSORB); HP(1); MaxHP(TEST_MAX_HP); };
        OPPONENT(SPECIES_WOBBUFFET);
    } WHEN {
        TURN { MOVE(opponent, MOVE_THUNDER_SHOCK); }
    } SCENE {
        ABILITY_POPUP(player, ABILITY_VOLT_ABSORB);
        HP_BAR(player, hp: TEST_MAX_HP / 4 + 1);
        MESSAGE("Jolteon restored HP using its Volt Absorb!");
    }
}
```

Reply:
Also, just to say, you'd want a `THEN` not a `FINALLY`. Currently `FINALLY` is silently ignored unless the test has a `PARAMETRIZE`, and #2955 will add a check for that.

Reply:
I have the worry that Water Absorb might have the same issue. I think we need to do a manual test.

Reply:
Can confirm it's working correctly in-game.

![multihittest2](https://user-images.githubusercontent.com/85393060/235565715-264cbe7c-1502-491b-a36b-7a515b47c8f6.gif)

Reply:
Ah ok, then it should be fine 👍